internal/cluster/invocation_factory.go

```diff
@@ -38,36 +38,36 @@ func NewConnectionInvocationFactory(config *pubcluster.Config) *ConnectionInvoca
 	}
 }
 
-func (f *ConnectionInvocationFactory) NewInvocationOnPartitionOwner(message *proto.ClientMessage, partitionID int32) *invocation.Impl {
+func (f *ConnectionInvocationFactory) NewInvocationOnPartitionOwner(message *proto.ClientMessage, partitionID int32, start time.Time) *invocation.Impl {
 	message.SetCorrelationID(f.makeCorrelationID())
-	return invocation.NewImpl(message, partitionID, "", time.Now().Add(f.invocationTimeout), f.redoOperation)
+	return invocation.NewImpl(message, partitionID, "", start.Add(f.invocationTimeout), f.redoOperation)
 }
 
-func (f *ConnectionInvocationFactory) NewInvocationOnRandomTarget(message *proto.ClientMessage, handler proto.ClientMessageHandler) *invocation.Impl {
+func (f *ConnectionInvocationFactory) NewInvocationOnRandomTarget(message *proto.ClientMessage, handler proto.ClientMessageHandler, start time.Time) *invocation.Impl {
 	message.SetCorrelationID(f.makeCorrelationID())
-	inv := invocation.NewImpl(message, -1, "", time.Now().Add(f.invocationTimeout), f.redoOperation)
+	inv := invocation.NewImpl(message, -1, "", start.Add(f.invocationTimeout), f.redoOperation)
 	inv.SetEventHandler(handler)
 	return inv
 }
 
-func (f *ConnectionInvocationFactory) NewInvocationOnTarget(message *proto.ClientMessage, addr pubcluster.Address) *invocation.Impl {
+func (f *ConnectionInvocationFactory) NewInvocationOnTarget(message *proto.ClientMessage, addr pubcluster.Address, start time.Time) *invocation.Impl {
 	message.SetCorrelationID(f.makeCorrelationID())
-	inv := invocation.NewImpl(message, -1, addr, time.Now().Add(f.invocationTimeout), f.redoOperation)
+	inv := invocation.NewImpl(message, -1, addr, start.Add(f.invocationTimeout), f.redoOperation)
 	return inv
 }
 
-func (f *ConnectionInvocationFactory) NewConnectionBoundInvocation(message *proto.ClientMessage, conn *Connection, handler proto.ClientMessageHandler) *ConnectionBoundInvocation {
+func (f *ConnectionInvocationFactory) NewConnectionBoundInvocation(message *proto.ClientMessage, conn *Connection, handler proto.ClientMessageHandler, start time.Time) *ConnectionBoundInvocation {
```

Comment:
The old way looks more convenient. Every caller of the API seems to be calling time.Now()
What was the reason of taking `start` as parameter ?

Reply:
Invocations are not mutated once created, so for each attempt a new invocation should be created (the message is copied). So, the deadline should be calculated at the time of the first invocation creation. (Otherwise the deadline would apply to a particular invocation, not the put/get, etc. request).

Reply:
We discussed with @yuce that we might use CircutirBreaker for this purpose, instead of passing time around internals. It will be addressed in another pr if it looks like a good fit.